perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_def_mfaEnforceSettings.java

```diff
@@ -0,0 +1,188 @@
+	 *
+	 * @param node JsonNode
+	 * @throws WrongAttributeValueException
+	 */
+	private void isStringArrayNode(JsonNode node, String name) throws WrongAttributeValueException {
+		// Check property is valid array
+		if (node.isArray()) {
+			// Check all items of array are string like
+			for (Iterator<JsonNode> it = node.elements(); it.hasNext(); ) {
+				JsonNode value = it.next();
+				if (!value.isTextual()) {
+					throw new WrongAttributeValueException("Property '" + name + "' has non textual value " + value);
+				}
+			}
+		} else {
+			throw new WrongAttributeValueException("Property '" + name + "' is not an array.");
+		}
+	}
+	
+	/**
+	 * The following restrictions are placed on the attribute value:
+	 * {"include_categories":["str1","str2"]} str1, str2 is an existing key in the entityless attribute mfaCategories
+	 * {"include_categories":["str1","str2"],"exclude_rps":["rp1","rp2"]} str1, str2 is an existing key in the entityless attribute mfaCategories and rp1, rp2 must exist inside the category
+	 *
+	 * @param perunSession PerunSession
+	 * @param user User
+	 * @param attribute Attribute of the user.
+	 *
+	 * @throws WrongReferenceAttributeValueException
+	 * @throws WrongAttributeAssignmentException
+	 */
+	@Override
+	public void checkAttributeSemantics(PerunSessionImpl perunSession, User user, Attribute attribute) throws WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		String val = attribute.valueAsString();
+		if (val == null || val.isEmpty()) return;
+
+		Set<String> includeCategories = null;
+		Set<String> excludeRps = null;
+		try {
+			final ObjectMapper mapper = new ObjectMapper();
+			JsonNode root = mapper.readTree(val);
+
+			// Syntax check ensures correctness for 'all'
+			if (root.has("all")) return;
+
+			// Initialize included categories to check
+			includeCategories = new HashSet<>();
+			fillSet(root.get("include_categories"), includeCategories);
+
+			// Initialize excluded rps to check (optional)
+			excludeRps = new HashSet<>();
+			if (root.has("exclude_rps")) {
+				fillSet(root.get("exclude_rps"), excludeRps);
+			}
+		} catch (Exception e) {
+			throw new WrongAttributeAssignmentException("Attribute " + attribute + "is incorrectly assigned.");
+		}
+
+		Attribute mfaCategories = perunSession.getPerunBl().getAttributesManagerBl().getEntitylessAttributes(perunSession, "mfaCategories").get(0);
+		String mfaSettingsValue = mfaCategories.valueAsString();
+		try {
+			final ObjectMapper mapper = new ObjectMapper();
+			JsonNode mfaCategoriesNode = mapper.readTree(mfaSettingsValue).get("categories");
+
+			// Iterate through categories and check that all included categories exist
+			for (Iterator<Map.Entry<String, JsonNode>> catIt = mfaCategoriesNode.fields(); catIt.hasNext(); ) {
+				Map.Entry<String, JsonNode> catEntry = catIt.next();
+				includeCategories.remove(catEntry.getKey());
+
+				// Iterate through rps and check that all excluded rps exist
```

Comment:
We also want to check that the excluded rps is from one of the included categories (would not make sense to exclude an rps from a category that is not included). 99% sure about this, but check what the mfa API used in the profile app returns in this scenario